Gutenboarding: `data-store` reducer function signatures don't need to list which actions they handle

p-jackson opened the pull request:
In order to get type checking and code completion, the data-store reducers have been explicitly listing each of the actions they handle in their type signature. I think this (1) doesn't scale well as more cases are added to the switch statement (2) makes the reducer's type too specific.

A reducer should be able to handle any of a store's actions without outside observers being aware any changes that are made. By listing them in the type they're effectively part of the function's API, and changing the implementation could be a breaking API change (loosely speaking).

**Edit**: and converting the reducers to function expressions is entirely optional. I just did it to drive home that adding TypeScript to reducers doesn't require you to jump through any extra hoops, write reducers in a way that's different from JS reducers, or declare somethings type before defining it.
But I don't mean to bikeshed on a stylistic thing 😅 I made the change in a single commit so it's easy to drop.

#### Changes proposed in this Pull Request

Reducer action types are still generated automatically from the action creators, but now it includes all actions.

* ~Add a new `ActionsDefinedInModule` mapped type~
* Reducer function signatures include all actions for that store
* ~Switch the reducers to use function expressions (optional)~

p-jackson committed:
data-store reducers accept all applicable actions

ramonjd commented:
@p-jackson  

> the data-store reducers have been explicitly listing each of the actions they handle in their type signature. I think this (1) doesn't scale well as more cases are added to the switch statement (2) makes the reducer's type too specific.

I like the motivation behind this PR, and think it makes 100% sense. 

I don't feel confident reviewing it from a TS perspective yet, but it doesn't break the frontend and I can't see any compilation errors ☺️

p-jackson commented:
Thanks @ramonjd, my main worry was it didn't make sense. I was struggling to word my rationale 😅 

Obviously hoping for some 👀 from @Automattic/luna, but maybe I'll throw a @Automattic/type-review in for good measure.

andrewserong commented:
Nice work @p-jackson, I also don't feel confident reviewing it from a TS perspective, but for me this also greatly improves the readability of the reducers as that top block of passing in generics to `Reducer` took quite a while for me to parse visually.

In practical terms, I think it also makes sense because won't each of the reducers be called with each of the actions, whether or not their type is specified in the reducer's signature? So I think this change feels a bit more accurate to the behaviour of reducers.

> Edit: and converting the reducers to function expressions is entirely optional.

I'm quite partial to function definitions and function expressions when functions are declared at the top-level, so I like this, too!

jsnajdr commented on client/landing/gutenboarding/stores/onboard/reducer.ts:
```diff
@@ -1,62 +1,48 @@
+type DomainSuggestion = import('@automattic/data-stores').DomainSuggestions.DomainSuggestion;
+type Template = import('@automattic/data-stores').VerticalsTemplates.Template;
+
+function domain( state: DomainSuggestion | undefined = undefined, action: OnboardAction ) {
```
Setting the `state`'s default value to `undefined` is redundant.

Also, using the `Reducer<>` type has some benefits, as it adds `# | undefined` automatically to the state type:
```js
const domain: Reducer< DomainSuggestion, OnboardAction > = ( state, action ) => {
  ...
}
```

I continue to be very confused about the `OnboardAction` type. I believe that the action type shouldn't be constrained in any way beyond requiring an object with `type` field. And that's what the `import { Action } from 'redux'` type specifies.

Redux will call the reducer with an init action like this one:
```js
{
  type: "@@redux/INITj.3.e.w.g"
}
```
This action doesn't pass typecheck for `OnboardingAction`, and yet the reducer is guaranteed to be called with it and must react to it correctly.

On the other hand, I see exactly this kind of constraining the action type in the official docs, too:
https://redux.js.org/recipes/usage-with-typescript/#type-checking-reducers
Is it a bug, or am I just clueless about something?

sirreal replied:
The `Reducer` type is very strange and I've wondered about this myself. We _claim_ that a `Reducer` handles a specific action type, but as you say it must handle _all_ actions.

I'd speculate it's a tradeoff the authors decided to make because by _convention_ —you _can't_ use Redux if you don't understand the `if ( action.type === 'MY_ACTION' ) { return 'new state' } else { return state }` pattern—, reducers gracefully handle all action types.

If we expect this by convention and lie a little bit to the type system, we get better information in our reducer functions. A correctly typed reducer function that wants to deal with a specific action but has to deal with actions of `{ type: any }` ends up being painful to work with. Consider that we can't even get a discriminated union because the basic action type `{ type: any }` also satisfies a fully typed `{ type: 'MY_ACTION', foo: string, bar: number, baz: boolean }`, so in our `if ( action.type === 'MY_ACTION' ) { … }` branch, we still don't know whether which type we have!

I think the following illustrates it:

```ts
import { Reducer } from 'redux'

interface ReducerCorrect<S, A> {
    // Note the `A | A_` here, our type effectively opens up to `{ type: any }` 😞
    <A_ extends { type: any; }>(state: S | undefined, action: A | A_): S;
}


type Action = { type: 'foo', bar: { baz: { quux: number } } }

// This is unsafe, but the experience is better
const reducerRedux: Reducer<number, Action> = (state = 0, action) => {
     // Whoa, that's going to error at runtime for sure!
    // Of course, anyone experienced with Redux won't write this…
    return action.bar.baz.quux;
}

// This is safe, but is a pain
const reducerCorrect: ReducerCorrect<number, Action> = (state = 0, action) => {
    if (action.type === 'MY_ACTION') {
        return action.bar.baz.quux; // We can't know this -- type error! 😭
    } 
    return state;
}
```

p-jackson replied:
My thoughts about typing tradeoffs echo @sirreal's. The global nature of Redux actions almost makes them untypable. Elm's pattern of using nested actions is a way of making all the types perfect, but global events are pretty handy :)

> Setting the state's default value to undefined is redundant.

Good point, I'll clean that up while I'm here.

> using the Reducer<> type has some benefits, as it adds # | undefined automatically to the state type

I actually like that not all state types are `# | undefined`. Yes the onboard store has a lot of optional pieces of state so it'd be a convenience here. But the `domainSuggestion` state is never `undefined` and I feel like a lot of reducers work this way e.g. a lot of reducers are just a single value type and have a default.

sirreal replied:
🤔After considering this a bit more and seeing some other context (p1579980641021900-slack-typescript), I'd change my speculative stance 🙂 

The second type argument to `Reducer` is fine if left to the default. But, if it becomes the union of the type of _all actions in your application_, it should give excellent information and type safety.

p-jackson replied:
> Setting the `state`'s default value to `undefined` is redundant.

I've pushed a fix for this.

> The second type argument to `Reducer` is fine if left to the default. But, if it becomes the union of the type of all actions in your application, it should give excellent information and type safety.
 
@sirreal Just to clarify, does that mean you're happy with the changes here since it allows reducers to accept any of the actions for a given store? I thought that was your original (speculative) stance.

sirreal replied:
> does that mean you're happy with the changes […]

I've been considering these changes and playing with the types but haven't found the time to leave a full review. I'd prefer to leave it for a while.

jsnajdr replied:
It seems that the TypeScript types for Redux reducers are buggy. When I created a toy example of a reducer that accepts only a constrained set of actions, it indeed fails to compile valid dispatches of unknown actions:
```js
function reducer( state, action: HandledActionTypes ) {
  ...
}
reducer( undefined, { type: '@@INIT' } );
createStore( reducer ).dispatch( { type: '@@INIT' } );
```
Both these statements fail to compile and compiler reports an error:
```
Type '"@@INIT"' is not assignable to type '"ACTION_A" | "ACTION_B"'
```
The Redux repo already has a similar issue: https://github.com/reduxjs/redux/issues/3580
It describes another funny consequence of the constrained action type: one can write a reducer that fails to handle unknown actions:
```js
function reducer( state: MyState = initState, action: MyActions ): MyState {
  switch ( action.type ) {
    case MY_ACTION_A:
      return { ...state, ... };
    case MY_ACTION_B:
      return { ...state, ... };
  }
}
```
The switch statement exhausts all possible values of `action.type` for the `MyActions` type, so TypeScript doesn't mind that it returns `undefined` for all other `action.type`s. It thinks it can never happen 🙂 But such a reducer breaks the Redux contract, and also fails to always return a value of type `MyState`.

I added a comment that describes our case: https://github.com/reduxjs/redux/issues/3580#issuecomment-579239622

p-jackson replied:
Thanks for pointing out that Redux issue thread @jsnajdr, we're not the only one's grappling with how Redux and TS should work together.

I took a look at the [solution Simplenote is using](https://github.com/Automattic/simplenote-electron/blob/develop/CONTRIBUTING.md#redux-and-typescript), and their typings suffer from the same issue i.e. it's possible to write a reducer with no type errors that doesn't deal with unknown action types.

This isn't ideal, but given it's something community hasn't figured out yet (the approach recommended by the docs has the issue) is this something where we just wait and see where Redux goes with this. This PR improves things a little at least by widening the action type.

p-jackson committed:
Remove redundant setting state's default value to undefined

sirreal commented:
I've continued to have higher priority things get in the way of coming back to this. I was hoping to leave detailed feedback and propose some changes and improvements, but I've bee

sirreal commented:
Rebased.

I accidentally posted the previous incomplete comment.

I've continued to have higher priority things get in the way of coming back to this. I was hoping to leave detailed feedback and propose some changes and improvements, but I've been unable to find the time.

In general I think there are opportunities for simplification around here. I'm not sure I like the ergonomics of the new mapped types. Needing to import the namespace `import * as actions from './actions'`, then pass the typeof that namespace object to the mapped type doesn't seem ideal. It might make more sense to export an `Action` type from the `actions` module that is a union of all the possible actions.

There also seems to be opportunity for improvement with the type of the type property of actions. I'd like to remove the action-type enums, which don't make sense with a small set of action creators.

I started exploring and working on some ideas which I've shared in https://github.com/Automattic/wp-calypso/pull/39259, but I haven't found the time to finish it up.

p-jackson commented:
Thanks for commenting @sirreal, appreciate that you're trying to balance priorities :)

fwiw I like removing the enum for action types, I don't really feel adding the namespacing adds any value.

Interesting idea about removing the mapped type and explicitly defining a union of all action types. One advantage I see is that if some hypothetical action had a complicated enough type then we might want to define it explicitly as its own interface, instead of infer it from the action creator's return type, then it could easily be included in the union type.

p-jackson committed:
Remove action type enums

The extra namespacing the `ActionType` enums provided aren't worth the
extra typing. It makes the already long identifiers even longer, and you
already get auto-complete in reducers because TypeScript knows what
values are possible when you type `action.type ===` in a reducer.

p-jackson commented:
Rebased to include the new site store. I've updated to take some ideas from @sirreal's branch:
- action types no longer use an `enum`
  - they don't add a lot of value, we already get autocomplete in reducers without them
  - identifiers are already long enough to read/type without the extra prefix
- no longer use a mapped type for all possible actions

sirreal commented on client/landing/gutenboarding/stores/onboard/reducer.ts:
```diff
@@ -1,113 +1,87 @@
+type DomainSuggestion = import('@automattic/data-stores').DomainSuggestions.DomainSuggestion;
+type Template = import('@automattic/data-stores').VerticalsTemplates.Template;
+
+function domain( state: DomainSuggestion | undefined, action: OnboardAction ) {
```
I like using the `Reducer` type because it ensures we return a state type. I'd say it's a bit more cumbersome here using `function` declarations:

```suggestion
function domain( state: DomainSuggestion | undefined, action: OnboardAction ): DomainSuggestion {
```

If we stay with `const` declarations and `Reducer`, we get the following which seems like a good place:

```suggestion
const domain: Reducer<DomainSuggestion | undefined, OnboardAction> = ( state, action ) => {
```

If we don't return the same type of state, that's a bug types can help detect.

p-jackson replied:
I've dropped the commit that converted the reducers to function expressons. Given all selectors and action creators are using the `const` way it doesn't make sense for the reducers to be any different.

> I'd say it's a bit more cumbersome here using function declarations

Given that I think it's less cumbersome this way goes to show it must be a preference thing, so better to just go with the flow of what's already there.

Then again I _did_ forget to specify the return types, so maybe you have a point 😅

sirreal replied:
I don't have much of a preference as long as we're getting well typed reducers (with return):

```ts
const r: Reducer< State, Action > = ( s = initialValue, a ) => s;
// or
function r_( s: State = initialValue, a: Action ): State { return s; }

sirreal commented on packages/data-stores/src/user/actions.ts:
```diff
@@ -2,49 +2,57 @@
 	} catch ( err ) {
 		return receiveNewUserFailed( err );
 	}
 }
```
When working on this, I was curious about this.

- There's an untyped literal action object that's yielded in the middle. An observation, not necessarily an issue.
- This is a generator, which uniquely types its yield and return types.

I'm not sure whether our reducers "see" yielded actions (presumably handled by controls?) or only returned actions. Maybe this is up to the handling control and not something we can generalize about.

p-jackson replied:
Just tried this and was surprised to see that the reducer _didn't_ see the `'CREATE_ACCOUNT'` action 🤔

andrewserong replied:
@p-jackson I ran into this when I was working on the Site data store, and it looks like the way the [redux-routine middleware](https://developer.wordpress.org/block-editor/packages/packages-redux-routine/) works is that if it matches a control against the name of an action, it [runs the control](https://github.com/WordPress/gutenberg/blob/5920ab8620e0a5ea6ed810e399dc42fd94d99abd/packages/redux-routine/src/runtime.js#L26) but _doesn't_ dispatch the action to pass it along to the reducer. For unhandled actions (actions that don't have a match in the controls), they [do get dispatched](https://github.com/WordPress/gutenberg/blob/5920ab8620e0a5ea6ed810e399dc42fd94d99abd/packages/redux-routine/src/runtime.js#L40) by the middleware so that they hit the reducer as normal.

I still don't think I have a solid grasp of exactly how the middleware works, but in effect, it feels like controls are designed to intercept an action, and then in the control we can then dispatch actions that will hit the reducer. Because the reducer doesn't see the action that's listed in our controls, that's why we've got the additional `fetchNewUser` action that gets yielded before `CREATE_ACCOUNT`, otherwise from the reducer, we don't know that the action has started. It seems like useful behaviour for a middleware to me, but definitely non-obvious in its behaviour!

sirreal reviewed (approved):
Overall I'm quite happy with the direction, thanks for the initiative!

As noted, I do see reason to continue using `Reducer` or add explicit return types to the reducer `function` declarations. What do you think?

I've tested and there don't seem to be any regressions. Approving with the note that I'd like to see the above change 🚀

andrewserong commented on client/landing/gutenboarding/stores/onboard/actions.ts:
```diff
@@ -6,53 +6,65 @@ import { VerticalsTemplates } from '@automattic/data-stores';
 	domain: import('@automattic/data-stores').DomainSuggestions.DomainSuggestion | undefined
 ) => ( {
-	type: ActionType.SET_DOMAIN as const,
+	type: 'SET_DOMAIN' as const,
```
This is just a question for my own curiosity (and wrapping my head around TS features), what's the value of adding `as const` on a string literal? Does this prevent the `type` key from being changed?

p-jackson replied:
Without it the `type` prop would still be a `string`, not a string literal unfortunately.

Another interesting way of doing it is:
```typescript
return <const>{
	type: 'SET_DOMAIN',
	domain,
};
```

Which not only makes typeof `type` the correct string literal, but it also marks all the props as `readonly`.

Hover your mouse over the types in [this playground example](http://www.typescriptlang.org/play/?ssl=14&ssc=1&pln=15&pc=1#code/MYewdgzgLgBAhsKBLcBBGBeGBvAUDAmKATwAcBTALhgHJUBhAFQEkB5AOQH1Ub4IZQkKLgC+Ably4A9FJiMAFuSJklIAGZF5SftpgADOkzZcee3AmRoAdCQoTJkwdHiIUYAEKYc+Qraq0GFg5OdxpRexk5RWUKGHVNXV09aAAnJDAAczMLN3cbFXsHJ1gc8HovAB5igD48Qhj-QyCuejDxSUiFJT84jSgtHX4DQONOVuzXMvy7XCA) to see it in action.

andrewserong replied:
Ah, gotcha! Thanks for example, that makes it clear to me now :)

sirreal replied:
I'd like the reducers to all `return { … } as const` if anyone wants to take that task on 👍

andrewserong reviewed (approved):
This PR's looking great to me @p-jackson — thanks so much for sticking with it and rebasing against the Site data store, too, it makes the reducers much more readable to me while also better reflecting how the reducers function! 👍
